dask_cuda/explicit_comms/dataframe/shuffle.py

```diff
@@ -254,94 +222,54 @@ def shuffle(
     if npartitions is None:
         npartitions = df.npartitions
 
-    # Step (a): partition/group each dataframe-partition
+    # Step (a):
+    df = df.persist()  # Make sure optimizations are apply on the existing graph
+    wait(df)  # Make sure all keys has been materialized on workers
     name = (
-        "explicit-comms-shuffle-group-"
+        "explicit-comms-shuffle-getitem-"
         f"{tokenize(df, column_names, npartitions, ignore_index)}"
     )
-    df = df.persist()  # Making sure optimizations are apply on the existing graph
-    dsk = dict(df.__dask_graph__())
-    output_keys = []
-    for input_key in df.__dask_keys__():
-        output_key = (name, input_key[1])
-        dsk[output_key] = (
-            shuffle_group,
-            input_key,
+
+    # Stage all keys of `df` on the workers and cancel them, which makes it possible
+    # for the shuffle to free memory as the partitions of `df` are consumed.
+    rank_to_inkeys = c.stage_keys(name=name, keys=df.__dask_keys__())
+    c.client.cancel(df)  # Notice, since `df` has been staged, nothing is freed here.
+
+    # Find the output partition IDs for each worker
+    ranks = sorted(rank_to_inkeys.keys())
+    div = npartitions // len(ranks)
+    rank_to_out_part_ids: Dict[int, List[int]] = {}  # rank -> [list of partition id]
+    for i, rank in enumerate(ranks):
+        rank_to_out_part_ids[rank] = list(range(div * i, div * (i + 1)))
+    for rank, i in zip(ranks, range(div * len(ranks), npartitions)):
+        rank_to_out_part_ids[rank].append(i)
```

Comment:
Given that these are contiguous sequences, is it not more comms-efficient to just communicate the extents?

Reply:
Agree, we should compute the IDs on-the-fly eventually. However, I like to keep it simple while we experiment with different approaches.